text/0000-named-arguments.md

```diff
@@ -0,0 +1,623 @@
+```
+
+This also improves type safety, since a `Size` can't be passed to a function expecting a `Point`. However, this kind of type safety is often not required and not worth the additional complexity and boilerplate.
+
+### The builder pattern
+
+The builder pattern is a pattern where an intermediate object containing all the required parameters is created using method calls. This object is then passed to the actual function, which is also a method of the intermediate object. For example:
+
+```rust
+SplitBuilder::new("hello world") // positional arguments go here
+    .split_at(' ')               // the following methods emulate named arguments
+    .limit(2)
+    .case_sensitive(true)
+    .finish()                    // the actual function
+```
+
+The obvious downside is that this requires a lot of boilerplate. To reduce boilerplate, there are crates with procedural macros, which allow creating builder types in a declarative way. One problem with this is that it adds dependencies and increases compile times; in debug builds, it also affects runtime performance.
+
+The other problem is that all arguments are optional (it doesn't prevent you from calling `SplitBuilder::new("hello world").finish()` directly), and this is not always desired. It is possible to create more intermediate types and transitions that ensure that all required arguments are specified, but this adds more boilerplate (if implemented manually) and complexity.
+
+For large APIs with a lot of functions, using the builder pattern everywhere is not feasible.
+
+### Function names
+
+One workaround to make function calls more expressive is to add more information to the function name. For example:
+
+```rust
+// this:
+fn split(string: &str, .at: char, .limit: usize, .case_sensitive: bool) {}
+
+// could be written as
+fn split_at_with_limit_and_case_sensitivity(
+    string: &str,
+    at: char,
+    limit: usize,
+    case_sensitive: bool,
+) {}
+```
+
+However, the version using named arguments is more readable, it reads almost like a sentence:
+
+```rust
+split("Hello, world!", .at = ' ', .limit = 2, .case_sensitive = true)
+// vs.
+split_at_with_limit_and_case_sensitivity("Hello, world!", ' ', 2, true)
+```
+
+### Being realistic
+
+Instead of looking at how code _could_ be written in carefully crafted APIs, we should look at how code is being written _in reality_. Programmers don't always have time to rack their brains over how to create the most beautiful API. They want to get things done.
+
+Named arguments allow iterating quickly without sacrificing readability, because they are dead simple. There's no need to create new types or make up long function names.
+
+This is particularly true for private functions. Since they are not part of the public API, little thought is often given to making them readable. This makes the code harder to maintain as the code base grows. Ideally, both public and private code should be readable and expressive.
```

Comment:
This like this part anticipates opposition and paints it as _unrealistic_, while some of the patterns described before it are indeed used, some could even be considered common. It also makes statements like "dead simple", while adding a set of rules to how arguments work.

It also argues _against_ introducing new types, which I would generally not see as bad practice and find odd to find here.

The argument about private functions is similarly odd: I would like to see some better data that private functions are _indeed_ not written with similar care as public functions.